### ISSUE TITLE
rook-ceph only two monitors running, strange canary

### ISSUE THREAD

**Issue by ibotty** [rook/rook#5644]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**

There are only two monitors:
```
sh-4.2$ ceph -s
  cluster:
    id:     08f51f08-9551-488f-9419-787a7717555e
    health: HEALTH_OK
 
  services:
    mon: 2 daemons, quorum k,l (age 31h)
[...]
```
```
> oc get po -n rook-ceph | grep mon
rook-ceph-mon-k-56b79f9b8f-thmbz                                  1/1     Running       0          9d
rook-ceph-mon-k-canary-76975f59fb-sdc8k                           0/1     Pending       0          83s
rook-ceph-mon-l-6595d78777-9mhps                                  1/1     Running       0          9d
```

So there are indeed only two running.  The canary cannot run, because it cannot mount the RWO volume that is attached to `rook-ceph-mon-k-56b79f9b8f-thmbz`.

```
> oc get pvc
NAME              STATUS   VOLUME              CAPACITY   ACCESS MODES   STORAGECLASS   AGE
rook-ceph-mon-i   Bound    local-pv-5965673e   5Gi        RWO            local-nvme     9d
rook-ceph-mon-k   Bound    local-pv-304622ef   5Gi        RWO            local-nvme     9d
rook-ceph-mon-l   Bound    local-pv-e26b53e2   5Gi        RWO            local-nvme     9d
```

Interestingly, I did not get any alert, that there are only two monitors running.  That's a problem as well!

**Expected behavior:**

Three monitors running.


**File(s) to submit**:

* Cluster CR (custom resource), typically called `cluster.yaml`, if necessary
https://paste.centos.org/view/5ab7cce2

The relevant part might be

```
  mon:                                      
    count: 3                                                                                              
    allowMultiplePerNode: false                                                                           
    volumeClaimTemplate:               
      spec:                                                                                               
        storageClassName: local-nvme                                                                      
        resources:                                                                                                                                                                                                  
          requests:                                                                                                                                                                                                 
            storage: 5Gi                    

```

* Operator's logs, if necessary

https://paste.centos.org/view/90dc0742

* Crashing pod(s) logs, if necessary

 To get logs, use `kubectl -n <namespace> logs <pod name>`
When pasting logs, always surround them with backticks or use the `insert code` button from the Github UI.
Read [Github documentation if you need help](https://help.github.com/en/articles/creating-and-highlighting-code-blocks).

**Environment**:
This is on OKD 4.4.
```
> oc get clusterversion
version   4.4.0-0.okd-2020-05-23-055148-beta5   True        False         15d     Cluster version is 4.4.0-0.okd-2020-05-23-055148-beta5
```
* OS (e.g. from /etc/os-release):
* Kernel (e.g. `uname -a`):
* Cloud provider or hardware configuration:
* Rook version (use `rook version` inside of a Rook Pod):
```
sh-4.2$ rook version
rook: v1.3.5
go: go1.13.8
```

* Storage backend version (e.g. for ceph do `ceph -v`):

* Kubernetes version (use `kubectl version`): 
```
> oc version
Client Version: 4.5.0-202003312031-25645af
Server Version: 4.4.0-0.okd-2020-05-23-055148-beta5
Kubernetes Version: v1.17.1
```

* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): Openshift

**Comment by travisn** [comment]
@ibotty This is definitely unexpected. It looks like the operator is also failing over mons since you're already up to the letter L. What is the value of the `maxMonId` value in the configmap  `rook-ceph-mon-endpoints`? This value seems to be confused, as if it hasn't bee incremented since the last mon failover. It should be trying to start a new mon with the letter M instead of K.  The maxMonId is the number offset in the alphabet.

What could be causing all the mon failover? Have you had network or other issues in the cluster? Perhaps the bug is that the operator was restarted in the middle of the mon failover and the maxMonId wasn't updated as it should be. Just speculation so far though...

**Comment by ibotty** [comment]
The configmap  ` rook-ceph-mon-endpoints` is
```
  csi-cluster-config-json: '[{"clusterID":"rook-ceph","monitors":["172.30.75.90:6789","172.30.91.232:6789"]}]'
  data: l=172.30.91.232:6789,k=172.30.75.90:6789
  mapping: '{"node":{}}'
  maxMonId: "-1"
[...]
```

I had quiet some problems with overheating nodes so ceph was failing over mons for a while.  That did not happen for the last days though.  I don't know, why `mon-i`, which was happily running got deleted though.

**Comment by SerialVelocity** [comment]
I had the same problem yesterday where I only had two monitors running. The canaries were stuck in pending as there wasn't enough room left on the monitor hosts to schedule them.

It seems like there is an underlying issue when the canary pods go into pending state, the cluster can degrade.

(it would probably also be good to have the memory requests/limits removed for the canaries as they don't actually run anything and the monitors get recreated anyway)

**Comment by ibotty** [comment]
I am a little worried only having two monitors running.  How can I mitigate the problem?  What should the configmap contain?

**Comment by jbguerraz** [comment]
Facing same or similar issue here (this is not systematic, one cluster build will be fine, one won't)

2 mons are running.
Canary in pending but its mon pod is already here, up and running.
Canary is pending because anti-affinity rule disallow it to be scheduled since its mon is already there, using the slot.
Third mon won't appear since mons creation is sequential.

Deleting the pending canary related mon unlocked the situation.

**Comment by travisn** [comment]
> I am a little worried only having two monitors running. How can I mitigate the problem? What should the configmap contain?

@ibotty The maxMonId should be the numeric, zero-based index of the last running mon. In your case, `mon-l` was the last running mon so I would expect `maxMonId` to be `11`. Try updating it in the configmap and restart the operator.

> It seems like there is an underlying issue when the canary pods go into pending state, the cluster can degrade.

> (it would probably also be good to have the memory requests/limits removed for the canaries as they don't actually run anything and the monitors get recreated anyway)

@SerialVelocity The canaries are started with all the properties of the mon daemon that is desired to be scheduled in order to detect the scheduling decision from k8s. If the canary didn't have the same properties such as anti-affinity or resource limits, Rook wouldn't get an accurate scheduling decision from the canary pod. But if the canary pod is not able to be scheduled, Rook could be smarter about giving up on it sooner.

**Comment by ibotty** [comment]
Oh, I did what @jbguerraz mentioned, letting the canaries run.  A monitor started (`mon-d`‽) and got the following configmap.

```
data:
  csi-cluster-config-json: '[{"clusterID":"rook-ceph","monitors":["172.30.75.90:6789","172.30.137.16:6789","172.30.91.232:6789"]}]'
  data: k=172.30.75.90:6789,d=172.30.137.16:6789,l=172.30.91.232:6789
  mapping: '{"node":{"d":null,"k":null,"l":null}}'
  maxMonId: "3"
```

Is this going to be a problem going forward (because `k` and `l` are beyond `maxMonId=3`?

**Comment by travisn** [comment]
> Is this going to be a problem going forward (because `k` and `l` are beyond `maxMonId=3`?

In the short term it should be fine, but the operator will be confused in the future if it tries to failover to k or l and they already exist. If you go ahead and update the maxMoniId to 11, it will avoid this issue.

**Comment by ibotty** [comment]
I thought so.  Thank you for your help.

My problem is resolved, but I am a little alarmed that this happened.  What should I do?  Close this issue?  I will open a bug to track the alarm that was not raised with only two monitors running.

**Comment by travisn** [comment]
Let's still track this issue here, no need to open a new issue. It would be really helpful to get the operator logs if we hit a repro.

**Comment by ibotty** [comment]
Unfortunately I don't have the operator logs anymore.  I hope not to encounter the problem again.  But I will collect logs when it happens!

**Comment by ibotty** [comment]
Argh.  It happened again.

```
> oc get po
rook-ceph-mon-d-56bdbb5c9c-sm2k5                                  1/1     Running       0          22h
rook-ceph-mon-d-canary-f4cb5fb85-lllzh                            0/1     Pending       0          42s
rook-ceph-mon-l-6595d78777-4k8zj                                  1/1     Running       0          22h
> oc get configmaps  rook-ceph-mon-endpoints -o yaml
[...]
data:
  csi-cluster-config-json: '[{"clusterID":"rook-ceph","monitors":["172.30.137.16:6789","172.30.91.232:6789"]}]'
  data: d=172.30.137.16:6789,l=172.30.91.232:6789
  mapping: '{"node":{}}'
  maxMonId: "-1"
[...]
```

I collected the last two logs.  Unfortunately the previous operator did not have any logs anymore.

https://paste.centos.org/view/d0952b41

I will reset the maxMonId again.

**Comment by travisn** [comment]
@ibotty I found a fix for the operator to automatically update the maxMonId if it is not correct. However, I am not seeing yet why the maxMonId would be incorrect in the first place.

**Comment by travisn** [comment]
#5698 will fix what was possibly the root cause for the maxMonId from not being updated in the first place. If there was some issue removing the failed mon, the maxMonId was not being updated.

**Comment by ribx** [comment]
@travisn we have a similar problem, this is still on 1.3.7. Should this problem also be fixed in 1.3.8?

here is the operator log:
```2020-08-05T15:33:33.820842100+02:00 2020-08-05 13:33:33.820802 I | op-mon: adding mons. currently 2 mons are in quorum and the desired count is 3.
2020-08-05T15:33:33.930866097+02:00 2020-08-05 13:33:33.930825 I | op-mon: sched-mon: created canary deployment rook-ceph-mon-b-canary
2020-08-05T15:33:33.965061369+02:00 2020-08-05 13:33:33.965033 I | op-mon: sched-mon: waiting for canary pod creation rook-ceph-mon-b-canary
2020-08-05T15:33:39.746721683+02:00 2020-08-05 13:33:39.746685 I | op-cluster: detecting the ceph image version for image ceph/ceph:v14.2.9...
2020-08-05T15:33:41.798078418+02:00 2020-08-05 13:33:41.798040 I | op-cluster: Detected ceph image version: "14.2.9-0 nautilus"
2020-08-05T15:33:41.873891633+02:00 2020-08-05 13:33:41.873870 I | op-mon: parsing mon endpoints: c=10.97.43.152:6789,b=10.111.178.14:6789
2020-08-05T15:33:41.874053627+02:00 2020-08-05 13:33:41.874038 I | cephconfig: writing config file /var/lib/rook/rook-ceph/rook-ceph.config
2020-08-05T15:33:41.874107088+02:00 2020-08-05 13:33:41.874095 I | cephconfig: generated admin config in /var/lib/rook/rook-ceph
2020-08-05T15:33:42.203856739+02:00 2020-08-05 13:33:42.203816 I | op-cluster: cluster "rook-ceph": version "14.2.9-0 nautilus" detected for image "ceph/ceph:v14.2.9"
2020-08-05T15:33:42.298039209+02:00 2020-08-05 13:33:42.298017 I | op-config: CephCluster "rook-ceph" status: "Progressing". "Cluster is checking if updates are needed"
2020-08-05T15:36:05.024564347+02:00 2020-08-05 13:36:05.024509 I | op-mon: cleaning up canary monitor deployment "rook-ceph-mon-b-canary" and canary pvc "".
2020-08-05T15:36:05.024564347+02:00 2020-08-05 13:36:05.024529 I | op-k8sutil: removing deployment rook-ceph-mon-b-canary if it exists
2020-08-05T15:36:05.079649580+02:00 2020-08-05 13:36:05.079625 I | op-k8sutil: Removed deployment rook-ceph-mon-b-canary
2020-08-05T15:36:05.106484507+02:00 2020-08-05 13:36:05.106462 I | op-k8sutil: "rook-ceph-mon-b-canary" still found. waiting...
2020-08-05T15:36:07.124581624+02:00 2020-08-05 13:36:07.124546 I | op-k8sutil: confirmed rook-ceph-mon-b-canary does not exist
2020-08-05T15:36:07.124627951+02:00 2020-08-05 13:36:07.124584 W | op-mon: failed to check mon health. failed to assign pods to mons: assignmon: error scheduling monitor: sched-mon: canary pod scheduling failed retries
2020-08-05T15:36:07.124627951+02:00 2020-08-05 13:36:07.124594 I | op-mon: start running mons
2020-08-05T15:36:07.164581080+02:00 2020-08-05 13:36:07.164559 I | op-mon: parsing mon endpoints: c=10.97.43.152:6789,b=10.111.178.14:6789
2020-08-05T15:36:07.253094655+02:00 2020-08-05 13:36:07.253072 I | op-mon: saved mon endpoints to config map map[csi-cluster-config-json:[{"clusterID":"rook-ceph","monitors":["10.97.43.152:6789","10.111.178.14:6789"]}] data:b=10.111.178.14:6789,c=10.97.43.152:6789 mapping:{"node":{}} maxMonId:2]
2020-08-05T15:36:07.289776949+02:00 2020-08-05 13:36:07.289737 I | cephconfig: writing config file /var/lib/rook/rook-ceph/rook-ceph.config
2020-08-05T15:36:07.289812195+02:00 2020-08-05 13:36:07.289804 I | cephconfig: generated admin config in /var/lib/rook/rook-ceph
2020-08-05T15:36:07.561143215+02:00 2020-08-05 13:36:07.561114 I | op-mon: targeting the mon count 3
2020-08-05T15:36:07.847775357+02:00 2020-08-05 13:36:07.847721 I | op-mon: sched-mon: created canary deployment rook-ceph-mon-c-canary
2020-08-05T15:38:39.068508306+02:00 2020-08-05 13:38:39.068464 I | op-mon: cleaning up canary monitor deployment "rook-ceph-mon-c-canary" and canary pvc "".
2020-08-05T15:38:39.068508306+02:00 2020-08-05 13:38:39.068484 I | op-k8sutil: removing deployment rook-ceph-mon-c-canary if it exists
2020-08-05T15:38:39.118723028+02:00 2020-08-05 13:38:39.118698 I | op-k8sutil: Removed deployment rook-ceph-mon-c-canary
2020-08-05T15:38:39.148261213+02:00 2020-08-05 13:38:39.148240 I | op-k8sutil: "rook-ceph-mon-c-canary" still found. waiting...
2020-08-05T15:38:41.179783205+02:00 2020-08-05 13:38:41.179743 I | op-k8sutil: confirmed rook-ceph-mon-c-canary does not exist
2020-08-05T15:38:41.179827528+02:00 2020-08-05 13:38:41.179813 E | op-cluster: failed to create cluster in namespace "rook-ceph". failed to start the mons: failed to assign pods to mons: assignmon: error scheduling monitor: sched-mon: canary pod scheduling failed retries
2020-08-05T15:38:45.746719827+02:00 2020-08-05 13:38:45.746676 I | op-cluster: detecting the ceph image version for image ceph/ceph:v14.2.9...
2020-08-05T15:38:48.070777138+02:00 2020-08-05 13:38:48.070737 I | op-cluster: Detected ceph image version: "14.2.9-0 nautilus"
2020-08-05T15:38:48.146054200+02:00 2020-08-05 13:38:48.146016 I | op-mon: parsing mon endpoints: b=10.111.178.14:6789,c=10.97.43.152:6789
2020-08-05T15:38:48.146240099+02:00 2020-08-05 13:38:48.146226 I | cephconfig: writing config file /var/lib/rook/rook-ceph/rook-ceph.config
2020-08-05T15:38:48.146306544+02:00 2020-08-05 13:38:48.146294 I | cephconfig: generated admin config in /var/lib/rook/rook-ceph
2020-08-05T15:38:48.449506079+02:00 2020-08-05 13:38:48.449468 I | op-cluster: cluster "rook-ceph": version "14.2.9-0 nautilus" detected for image "ceph/ceph:v14.2.9"
2020-08-05T15:38:48.576085548+02:00 2020-08-05 13:38:48.576057 I | op-config: CephCluster "rook-ceph" status: "Progressing". "Cluster is checking if updates are needed"
```

and the config map:
```
data:
  csi-cluster-config-json: >-
    [{"clusterID":"rook-ceph","monitors":["10.97.43.152:6789","10.111.178.14:6789"]}]
  data: 'b=10.111.178.14:6789,c=10.97.43.152:6789'
  mapping: '{"node":{}}'
  maxMonId: '2'
```

The problem is, the operator tries to start canaries for the existing monitors and these canaries can never be scheduled, as the PVs are already bound.

Event log says:
```
0/7 nodes are available: 1 node(s) didn't match pod affinity/anti-affinity, 1 node(s) didn't satisfy existing 
pods anti-affinity rules, 3 node(s) had taint {node-role.kubernetes.io/master: }, that the pod didn't 
tolerate, 3 node(s) had volume node affinity conflict.
```

**Comment by travisn** [comment]
Actually the fix will come out with v1.3.9 (see #5868), planning on releasing it later today.
